### PR TITLE
Fix disabled REST alerts tests

### PR DESCRIPTION
**Pull request by machadovilaca**
### What this PR does
Before this PR:

VirtControllerRESTErrorsBurst, VirtControllerRESTErrorsHigh, VirtHandlerRESTErrorsBurst and VirtHandlerRESTErrorsHigh were disabled due to flakiness.

After this PR:

Fix the flakiness and reenable the tests

Fixes #

jira-ticket: `https://issues.redhat.com/browse/CNV-40856`

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/testsuite/kubevirtresource.go [hunk 1]
```diff
@@ -198,8 +198,10 @@ func UpdateKubeVirtConfigValue(kvConfig v1.KubeVirtConfiguration) *v1.KubeVirt {
 	patch, err := strategicpatch.CreateTwoWayMergePatch(old, newJson, kv)
 	Expect(err).ToNot(HaveOccurred())
 
-	kv, err = virtClient.KubeVirt(kv.Namespace).Patch(context.Background(), kv.GetName(), types.MergePatchType, patch, metav1.PatchOptions{})
-	Expect(err).ToNot(HaveOccurred())
+	Eventually(func() error {
```
Comment (EdDev) [inline tests/testsuite/kubevirtresource.go:201]: Why is this needed?

**Review by EdDev — CHANGES_REQUESTED** [review]
There is a lot of changes happening in this PR.
Please try to express the changes reasoning and summarize what is done to solve it in the commit message.

Changes that have implications on other tests are preferred to be introduced with a dedicated commit so it can explain the overall reasoning.

File: tests/testsuite/kubevirtresource.go (hunk 1 above)
Comment (machadovilaca) [reply tests/testsuite/kubevirtresource.go:201]: `Context("Errors metrics"` is a destructive test that prevents `virt-operator` from working properly to test related alerts. Sometimes this patch call can fail while `virt-operator` is not ready yet.

Comment (EdDev) [reply tests/testsuite/kubevirtresource.go:201]: What you are saying here is that due to some test, we need to retry here for this specific thing to succeed? I find it too odd and frankly worrying.

I see two options here:
- Convert this `UpdateKubeVirtConfigValue` helper to return errors instead of asserting internally. This may be harder than it looks to handle.
- At the caller of the problematic test, wait (using eventually) for something else to stabilize before proceeding to call this helper.

File: tests/monitoring/component_monitoring.go [hunk 2]
```diff
@@ -161,34 +161,42 @@ var _ = Describe("[Serial][sig-monitoring]Component Monitoring", Serial, decorat
 	})
 
 	Context("Errors metrics", func() {
```
Comment (RamLavi) [inline tests/monitoring/component_monitoring.go:163]: in `Fix flakiness and reenable errors metrics tests` commit - can you elaborate how this flakiness is fixed in this commit?

Comment (machadovilaca) [reply tests/monitoring/component_monitoring.go:163]: Due to the destructive nature of the tests, sometimes in the cleanup phase of each test, the config had different versions, the simple step of reverting the deleted CRBs before restoring `virt-operator` of being scaled to 0, ensures the components go back to a healthy state much faster.

Comment (RamLavi) [reply tests/monitoring/component_monitoring.go:163]: sounds logical. Can you add it to the commit desc?

File: tests/testsuite/kubevirtresource.go (hunk 1 above)
Comment (machadovilaca) [reply tests/testsuite/kubevirtresource.go]: I was not able to understand why 

```
12:32:01:   [FAILED] Timed out after 10.001s.
12:32:01:   Unexpected error:
12:32:01:       <*errors.errorString | 0xc0017779a0>: 
12:32:01:       failed to call healthz endpoint: Get "https://localhost:10047//healthz": EOF, component: "virt-controller", pod: "virt-controller-785cf77dbf-87488"
12:32:01:       {
12:32:01:           s: "failed to call healthz endpoint: Get \"https://localhost:10047//healthz\": EOF, component: \"virt-controller\", pod: \"virt-controller-785cf77dbf-87488\"",
12:32:01:       }
12:32:01:   occurred
12:32:01:   In [AfterEach] at: tests/tests_suite_test.go:168 @ 05/23/24 12:32:01.467
12:32:01: 
12:32:01:   Full Stack Trace
12:32:01:     kubevirt.io/kubevirt/tests_test.resetToDefaultConfig()
12:32:01:     	tests/tests_suite_test.go:168 +0x98
12:32:01:     kubevirt.io/kubevirt/tests_test.testCleanup()
12:32:01:     	tests/tests_suite_test.go:154 +0x6a
12:32:01:     kubevirt.io/kubevirt/tests_test.init.func19()
12:32:01:     	tests/tests_suite_test.go:110 +0xf
```

is happening without this improvement. Any ideas?

@EdDev

File: tests/monitoring/component_monitoring.go (hunk 2 above)
Comment (machadovilaca) [reply tests/monitoring/component_monitoring.go:163]: added

File: tests/testsuite/kubevirtresource.go (hunk 1 above)
Comment (EdDev) [reply tests/testsuite/kubevirtresource.go:201]: Ohh my, based on the trace, this occurs on teardown of the whole suite [1].
When the Kubevir CR is updated, there is a mechanism in-place [2] to check it was "seen" by all other components (I find these kind of hack bad, but this is what we have :/ ). That mechanism uses the `healthz` endpoint of each component.
What we see here is that it fails to query  the virt-controller.

But this is not what you fixed now. You retried the patching, and that is not what failed from the trace you provided now.

[1] https://github.com/kubevirt/kubevirt/blob/4c9a05e593c0eb52aa6a2c12dbbb331ad52843cc/tests/tests_suite_test.go#L168
[2] https://github.com/kubevirt/kubevirt/blob/ce48d715971bcd2808c15823ea0dc329f481bee3/tests/utils.go#L707

Comment (machadovilaca) [reply tests/testsuite/kubevirtresource.go:201]: But I only saw this error happening without this eventually retries I think
I can try to validate the endpoint is working in the test itself, and see what happens

Comment (machadovilaca) [reply tests/testsuite/kubevirtresource.go]: Locally it seems that increasing the waiting period in the endpoint checks is enough

**Comment by machadovilaca** [comment]
retesting to see if succeeds again

/test pull-kubevirt-e2e-k8s-1.29-sig-monitoring

**Comment by machadovilaca** [comment]
@EdDev @RamLavi can you take another look?

**Comment by RamLavi** [comment]
Hey @machadovilaca 

Seems like `pull-kubevirt-e2e-k8s-1.30-sig-operator` lane is failing alot, can you check it out?

**Comment by machadovilaca** [comment]
> Hey @machadovilaca
> 
> Seems like `pull-kubevirt-e2e-k8s-1.30-sig-operator` lane is failing alot, can you check it out?

@RamLavi looking into it
but I see it is failing across multiple different PRs: https://prow.ci.kubevirt.io/?repo=kubevirt%2Fkubevirt&job=pull-kubevirt-e2e-k8s-1.30-sig-operator

**Comment by EdDev** [comment]
> > Hey @machadovilaca
> > Seems like `pull-kubevirt-e2e-k8s-1.30-sig-operator` lane is failing alot, can you check it out?
> 
> @RamLavi looking into it but I see it is failing across multiple different PRs: https://prow.ci.kubevirt.io/?repo=kubevirt%2Fkubevirt&job=pull-kubevirt-e2e-k8s-1.30-sig-operator

Yes, we have a problem for almost a week now, it is unlikely related.

File: tests/monitoring/component_monitoring.go [hunk 3]
```diff
@@ -206,7 +220,7 @@ var _ = Describe("[Serial][sig-monitoring]Component Monitoring", Serial, decorat
 
 		It("VirtOperatorRESTErrorsBurst and VirtOperatorRESTErrorsHigh should be triggered when requests to virt-operator are failing", func() {
 			scales.RestoreScale(virtOperator.deploymentName)
-			err = virtClient.RbacV1().ClusterRoleBindings().Delete(context.Background(), crb.Name, metav1.DeleteOptions{})
+			err = virtClient.RbacV1().ClusterRoleBindings().Delete(context.Background(), "kubevirt-operator", metav1.DeleteOptions{})
```
Comment (EdDev) [inline tests/monitoring/component_monitoring.go:223]: It is hard to understand from where this name comes from and its relation to other things in the test setup. Consider using a const and share it to the relevant places to clarify the relation.

File: tests/monitoring/component_monitoring.go [hunk 4]
```diff
@@ -161,34 +161,48 @@ var _ = Describe("[Serial][sig-monitoring]Component Monitoring", Serial, decorat
...
-			_, err = virtClient.RbacV1().ClusterRoleBindings().Create(context.Background(), crb, metav1.CreateOptions{})
-			if !errors.IsAlreadyExists(err) {
-				util.PanicOnError(err)
+			for _, crb := range crbBackups {
+				crb.Annotations = nil
+				crb.ObjectMeta.ResourceVersion = ""
+				crb.ObjectMeta.UID = ""
+				_, err = virtClient.RbacV1().ClusterRoleBindings().Create(context.Background(), crb, metav1.CreateOptions{})
+				if !errors.IsAlreadyExists(err) {
+					util.PanicOnError(err)
```
Comment (RamLavi) [inline tests/monitoring/component_monitoring.go:195]:
```suggestion
					Expect(err).ToNot(HaveOccurred())
```

File: tests/monitoring/component_monitoring.go [hunk 5]
```diff
@@ -161,34 +161,48 @@ var _ = Describe("[Serial][sig-monitoring]Component Monitoring", Serial, decorat
...
+				}
 			}
-			scales.RestoreAllScales()
 
-			time.Sleep(10 * time.Second)
```
Comment (RamLavi) [inline tests/monitoring/component_monitoring.go:196]: +1

**Review by RamLavi — CHANGES_REQUESTED** [review]

File: tests/monitoring/component_monitoring.go (hunk 4 above)
Comment (EdDev) [reply tests/monitoring/component_monitoring.go:195]: This portion has not changed from the existing. Doing such a change will require reasoning.

Comment (RamLavi) [reply tests/monitoring/component_monitoring.go:195]: won't insist. 
just making a little improvement on the way. 
Could be in a separate PR.

Comment (machadovilaca) [reply tests/monitoring/component_monitoring.go:195]: here I'm panicking because if this fails, the after test cleanups will not succeed either

File: tests/monitoring/component_monitoring.go (hunk 3 above)
Comment (machadovilaca) [reply tests/monitoring/component_monitoring.go]: updated

**Comment by machadovilaca** [comment]
updated @EdDev

File: tests/monitoring/component_monitoring.go (hunk 4 above)
Comment (EdDev) [reply tests/monitoring/component_monitoring.go:194]: Not sure I understand the reasoning.
asserts are also panicking under the hood and an explicit panic may not really skip teardown operations. If it does, it will be very bad.

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by machadovilaca** [comment]
/retest
since https://github.com/kubevirt/kubevirt/pull/12055 got merged

**Comment by machadovilaca** [comment]
@assafad connection metric seems to still fail here
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11950/pull-kubevirt-e2e-k8s-1.29-sig-monitoring/1800456410966790144
